cli/commands/dump/dump.go

```diff
@@ -0,0 +1,155 @@
+package dump
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/sensu/sensu-go/cli"
+	"github.com/sensu/sensu-go/cli/commands/flags"
+	"github.com/sensu/sensu-go/cli/commands/helpers"
+	"github.com/spf13/cobra"
+)
+
+const (
+	// VerbList is the sensuctl verb to list resources.
+	VerbList = "list"
+	// VerbInfo is the sensuctl verb to get info about a resource.
+	VerbInfo = "info"
+)
+
+// Action is a resource/verb tuple for sensuctl commands.
+type Action struct {
+	Resource   string
+	Verb       string
+	Namespaced bool
+}
+
+var (
+	// All is all the core resource types and associated sensuctl verbs (non-namespaced resources are intentionally ordered first).
+	All = []Action{
+		Action{Resource: "namespace", Verb: VerbList, Namespaced: false},
+		Action{Resource: "cluster-role", Verb: VerbList, Namespaced: false},
+		Action{Resource: "cluster-role-binding", Verb: VerbList, Namespaced: false},
+		Action{Resource: "user", Verb: VerbList, Namespaced: false},
+		Action{Resource: "tessen", Verb: VerbInfo, Namespaced: false},
+		Action{Resource: "asset", Verb: VerbList, Namespaced: true},
+		Action{Resource: "check", Verb: VerbList, Namespaced: true},
+		Action{Resource: "entity", Verb: VerbList, Namespaced: true},
+		Action{Resource: "event", Verb: VerbList, Namespaced: true},
+		Action{Resource: "filter", Verb: VerbList, Namespaced: true},
+		Action{Resource: "handler", Verb: VerbList, Namespaced: true},
+		Action{Resource: "hook", Verb: VerbList, Namespaced: true},
+		Action{Resource: "mutator", Verb: VerbList, Namespaced: true},
+		Action{Resource: "role", Verb: VerbList, Namespaced: true},
+		Action{Resource: "role-binding", Verb: VerbList, Namespaced: true},
+		Action{Resource: "silenced", Verb: VerbList, Namespaced: true},
+	}
+)
+
+// Command dumps generic Sensu resources to a file or STDOUT.
+func Command(cli *cli.SensuCli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "dump [RESOURCE TYPE],[RESOURCE TYPE]... [-f FILE]",
+		Short: "dump resources to a file or STDOUT",
+		RunE:  execute(cli),
+	}
+
+	helpers.AddFormatFlag(cmd.Flags())
+	helpers.AddAllNamespace(cmd.Flags())
+	_ = cmd.Flags().StringP("file", "f", "", "file to dump resources to")
+
+	return cmd
+}
+
+func execute(cli *cli.SensuCli) func(*cobra.Command, []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		if len(args) != 1 {
+			_ = cmd.Help()
+			return errors.New("invalid argument(s) received")
+		}
+
+		// get the configured format or the flag override
+		format := cli.Config.Format()
+		if flag := helpers.GetChangedStringValueFlag("format", cmd.Flags()); flag != "" {
+			format = flag
+		}
+		switch format {
+		case "yaml", "wrapped-json":
+		default:
+			format = "yaml"
```

Comment:
This bit is a bit counterintuitive for me. If you print the help usage, it tells you the following:

```
$ sensuctl dump -h
[...]
      --format string    format of data returned ("json"|"wrapped-json"|"tabular") (default "tabular")
```

Which is actually not true, since the default format is `yaml` and only `wrapped-json` & `yaml` are supported. I think we should remove the default format flag (added at the line 60 above) and create our own flag with the correct default & allowed values.

Reply:
Kinda stole this from https://github.com/sensu/sensu-go/blob/daa6748dfc5d808f65418b474a2fdb58b1ba40a4/cli/commands/edit/edit.go#L45-L48 but I suppose it would be a little confusing if someone did `--format tabular` and got yaml instead. I agree, I'll add a special flag rather than the global one.

Reply:
Actually, here is another weird UX thing. Say I'm configured in default tabular format, and run `sensuctl dump all` without a format flag. In other cases of sensuctl we would want to use the current format, however this flag isn't consistent with that. I'm curious if a format flag which defaults to yaml rather than what's configured would be confusing to users..

Reply:
> I'm curious if a format flag which defaults to yaml rather than what's configured would be confusing to users..

We don't have to change this behavior, I'm simply saying the help usage should be modified to reflect what's happening behind the scenes. We can simply avoid calling `helpers.AddFormatFlag(cmd.Flags())` and instead manually define the `format` flag so at least the help usage displays the right values as default and supported, while still trying the use the value the user has in its config.

Showing `tabular` or even `json` as a support format value for this command is more confusing to me, because it's not the reality.

Reply:
- Updated help usage in `dump` to read `--format string    format of data returned ("wrapped-json"|"yaml") (default "wrapped-json")` (where wrapped-json is the current config)
- Updated help usage in the global format flag to read `--format string    format of data returned ("json"|"wrapped-json"|"tabular"|"yaml") (default "tabular")` (yaml was missing before)